### ISSUE TITLE
Improve query tokenizer for query classification

### ISSUE THREAD

**Issue by skekre98** [skekre98/NBA-Search#23]
**Description of feature/enhancement**
This is the tokenizer that will be used in the classification pipeline. Currently using very standard tokenization by simply removing stop words and punctuation. 

**Description of implementation**
Improve `query_tokenizer` function in transformer module. This should be more oriented towards NBA queries and not generalized text classification. Recommend using `spacy` for feature extraction. Current implementation uses standard text cleaning without any aspect of sports like queries:
`def query_tokenizer(sentence):
    mytokens = parser(sentence)
    mytokens = [word.lemma_.lower().strip() if word.lemma_ != "-PRON-" else word.lower_ for word in mytokens]
    mytokens = [word for word in mytokens if word not in stopwords and word not in punctuations]
    return mytokens`

**Why this is a good feature/enhancement**
Need to improve the accuracy for query classification model.

**Additional context**
This model will be used to classify queries received from the user for the chatbot.

**Comment by Saif807380** [comment]
Can I work on this?

**Comment by skekre98** [comment]
@Saif807380 Yes, that would be much appreciated 👍

**Comment by Saif807380** [comment]
What does it mean to be "more oriented towards NBA queries"? 

Should the tokenizer not include tokens that are not related to the NBA? 

Also the query classification is to be done so that the queries will be better classified as rank queries or stat queries right?

**Comment by skekre98** [comment]
Not necessarily not related to the NBA, but currently the tokenizer is simply removing stop words and checking for POS such as "pronoun". A better tokenizer might check for multiple POS or look for something that better differentiates NBA queries. Such as a rank query usually has two proper nouns since multiple people are bring compared while a stat query tends to have one. 

Yes, for now this should be done so classification is easier for rank and stat queries. Down the road this classifier should be able to handle multi-label classification(just something to keep in back of mind 😄)

**Comment by Saif807380** [comment]
So here's what I think -

* Keep tokens like [what, where, which, who, how, why] as they are direct indicators for the two classes. For example, how indicates a stat query, who indicates a rank one.
* Don't remove pronouns as they will also remove some of the wh- tokens.
* Keep token "or" which is another direct indicator of rank.
* Keep adjectives as they help quantify tokens like points, etc.
* Keep named entities. No need to count the number of named entities as the job of the tokenizer is to simply extract tokens of importance and not to get tokens that help in differentiating between the 2 classes.
* Keep dates as they are direct indicators of stats.

**Comment by Saif807380** [comment]
Also is there a way to test the classifier as right now even without any changes it gives 100% accuracy on `query.csv`?

**Comment by skekre98** [comment]
Yes, this area needs a lot of improvement 😅. It was just a place holder while testing the Inference Network workflow. Currently #26 is addressing parts of this by making training data more uniform. I will definitely be opening more issues around this subject such as:
-  Create a script to test `query_classifier` and any other models that may be needed in the future
- Add another class label so we can address more queries. 

Definitely not limited to this, so let me know if you think of anything else 😊

**Comment by Saif807380** [comment]
Sure. One way to get more queries would be to store the queries made by the users themselves. It can be an excellent source of raw data better representing the actual questions the boy will receive.

**Comment by skekre98** [comment]
I've created #28, perhaps lets use that to discuss this in the future. For now I will close this issue.
